### PR TITLE
Implement KAS_PUT_BORROWS_ARRAY

### LINKED ISSUE #180: Add put method (or flag to open) where kastore neither copies or takes ownership of the buffer.
#179 Added the ability of callers to take ownership of arrays that have been created and read by kastore, thereby avoiding a copy.
I would be possible to do this in reverse and allow kastore to take ownership of arrays that are being written.

I can imagine this only being useful in tskit when the `tsk_table_collection` dump is the last thing done to that `tsk_table_collection`, as it would no longer own its arrays.

### PR DESCRIPTION
**Pull request by benjeffery**
Fixes #180 

Still needs docs so draft, had to break up the put operation a bit, but I think this makes sense.

### CODE REVIEW

File: c/kastore.h [hunk 1]
```diff
@@ -492,6 +495,9 @@ int kastore_puts_float64(
 
 /** @} */
 
+int kastore_bput(kastore_t *self, const char *key, size_t key_len, const void *array,
```
Comment (jeromekelleher) [inline c/kastore.h:498]: What's the advantage of this vs using the existing ``kastore_put`` method? We can pass a flag ``KAS_BORROW_ARRAY`` to it, right?

File: c/kastore.h [hunk 2]
```diff
@@ -94,8 +94,9 @@ Unknown flags were provided to open.
 /** @} */
 
 /* Flags for open */
-#define KAS_READ_ALL            (1 << 0)
+#define KAS_READ_ALL                       (1 << 0)
 #define KAS_GET_TAKES_OWNERSHIP            (1 << 1)
+#define KAS_PUT_BORROWS_ARRAY              (1 << 2)
```
Comment (jeromekelleher) [inline c/kastore.h:99]: I don't really see the advantage of making this a global open flag - why not pass a flag to kastore_put(...) each time? It already has the array argument as const, right, and the flags argument is currently unused.

File: c/kastore.c [hunk 3]
```diff
@@ -527,8 +535,13 @@ kastore_insert_all(kastore_t *self, kastore_t *other)
 
     for (j = 0; j < other->num_items; j++) {
         item = other->items[j];
-        ret = kastore_put(
-            self, item.key, item.key_len, item.array, item.array_len, item.type, 0);
+        if (other->flags && KAS_PUT_BORROWS_ARRAY) {
+            ret = kastore_put(
```
Comment (jeromekelleher) [inline c/kastore.c:539]: I suspect you meant to call bput here? I don't think it's worth worrying about this here as it's a static method, and only gets called in a very specific case anyway.

File: c/kastore.c [hunk 4]
```diff
@@ -362,9 +362,17 @@ kastore_write_data(kastore_t *self)
             goto out;
         }
         size = self->items[j].array_len * type_size(self->items[j].type);
-        if (size > 0 && fwrite(self->items[j].array, size, 1, self->file) != 1) {
-            ret = KAS_ERR_IO;
-            goto out;
+        if (self->flags & KAS_PUT_BORROWS_ARRAY) {
```
Comment (jeromekelleher) [inline c/kastore.c:365]: Can't we do something like this?
```c
const void *write_array;

item = self->items[j];
assert((item->borrowed_array == NULL) != (item->array == NULL));
write_array = item->borrowed_array != NULL? item->borrowed_array: item->array;
if (size > 0 && fwrite(write_array, size, 1, self->file) != 1) {
...
```
We shouldn't need to duplicate logic like this.

**Review by jeromekelleher — COMMENTED** [review]
Looks good. I agree we need to add the ``borrowed_array`` field, but I'm not sure about the global open-time flag and I don't think we need the ``bputs`` function.

**Comment by jeromekelleher** [comment]
Thinking a bit more, we definitely don't want a global open-time flag because we need to mix borrow and copy semantics on PUT. We do things like

```c
int some_function(store)
{
     int8_t version[2] = {0, 1};
     
     kastore_puts_int8(store, "version", version, 2...)
}
```
We can't have global borrow semantics on this because we must take a copy immediately within the scope of ``some_function``.

File: c/kastore.h (hunk 1 above)
Comment (benjeffery) [reply c/kastore.h:498]: kastore_put calls kastore_oput which takes a non-const array, so even with a flag we would still need something like `bput`.

Comment (jeromekelleher) [reply c/kastore.h:498]: That can be a static method though, not part of the public API, right?

**Comment by benjeffery** [comment]
@jeromekelleher Yep, this makes more sense. bput is now static and have moved the flag to put.

**Comment by benjeffery** [comment]
Flag check and changelog added - along with flag number change.

File: c/kastore.c [hunk 5]
```diff
@@ -919,8 +893,79 @@ kastore_oput(kastore_t *self, const char *key, size_t key_len, void *array,
...
+}
 
+int KAS_WARN_UNUSED
+kastore_oput(kastore_t *self, const char *key, size_t key_len, void *array,
+    size_t array_len, int type, int flags)
+{
+    int ret = 0;
+    kaitem_t *item;
+
+    if (flags != KAS_BORROWS_ARRAY && flags != 0) {
```
Comment (jeromekelleher) [inline c/kastore.c:957]: This doesn't make sense  - KAS_BORROWS_ARRAY is not a valid input to the "own_put" operation. We should error here if flags != 0.

This check should be in kastore_put.

File: c/kastore.h [hunk 6]
```diff
@@ -427,12 +432,14 @@ int kastore_gets_float64(
 @rst
 A key with the specified length is inserted into the store and associated with
 an array of the specified type and number of elements. The contents of the
-specified key and array are copied. Keys can be any sequence of bytes but must
-be at least one byte long and be unique. There is no restriction on the
-contents of arrays. This is the most general form of ``put`` operation in
-kastore; when the type of the array is known and the keys are standard C
-strings, it is usually more convenient to use the :ref:`typed variants
-<sec_c_api_typed_put>` of this function.
+specified key and array are copied unless KAS_BORROWS_ARRAY is specified.
```
Comment (jeromekelleher) [inline c/kastore.h:435]: might be more obvious to say "unless the KAS_BORROWS_ARRAY flag is specified" - there's no indiction of what KAS_BORROWS_ARRAY is here

File: c/tests.c [hunk 7]
```diff
@@ -730,6 +749,54 @@ test_take_ownership(void)
...
+
+    ret = kastore_open(&store, _tmp_file_name, "w", 0);
+    CU_ASSERT_EQUAL_FATAL(ret, 0);
+
+    ret = kastore_puts(&store, "c", array, 4, KAS_UINT32, 0);
+    CU_ASSERT_EQUAL_FATAL(ret, 0);
+    ret = kastore_puts(&store, "b", array, 2, KAS_UINT32, KAS_BORROWS_ARRAY);
+    CU_ASSERT_EQUAL_FATAL(ret, 0);
+    ret = kastore_puts(&store, "a", array, 1, KAS_UINT32, KAS_BORROWS_ARRAY);
+    CU_ASSERT_EQUAL_FATAL(ret, 0);
```
Comment (jeromekelleher) [inline c/tests.c:770]: We should add test for using one of the more specific forms too, to make sure the flags are passed through, e.g.kastore_puts_uint32

**Review by jeromekelleher — COMMENTED** [review]
LGTM, a few small issues.

**Comment by benjeffery** [comment]
@jeromekelleher comments addressed, thanks!

File: c/CHANGELOG.rst [hunk 8]
```diff
@@ -1,3 +1,11 @@
+--------------------
+[2.0.2] - 2022-01-XX
```
Comment (jeromekelleher) [inline c/CHANGELOG.rst:2]: This is adding a new feature, so I think it's a minor version bump rather than a patch bump?

**Review by jeromekelleher — APPROVED** [review]
LGTM, thanks!

I think we should be incrementing to version 2.1.0 though.

**Comment by benjeffery** [comment]
@jeromekelleher I optimistically put that we'll push a release of kastore today. Does that make sense?

**Comment by jeromekelleher** [comment]
Yep, sgtm!

**Pull request merged**